File: crates/normalize-languages/tests/fixtures/matlab/variants.m
% Completeness fixture: one construct per grammar-legal variant of each
% field the matlab.{tags,calls,imports,complexity,cfg,types}.scm queries
% constrain, cross-referenced against arborium-matlab 2.17.0's
% node-types.json. Every construct here is *expected to be captured*
% (except the NEGATIVE section); see query_fixtures.rs `matlab_*_completeness_*`
% tests for the matrix.
%
% NOTE: `.m` collides with Objective-C's extension in normalize's CLI, so
% this fixture is never resolved by extension guessing anywhere in the test
% harness — tests load it via `GrammarLoader::get("matlab")` explicitly (see
% query_fixtures.rs), the same mechanism sample.m's tests already use.

% --- function_call.name variants (matlab.calls.scm) -------------------------

function test_calls()
    plainCall(1); % name: identifier
    s.dynamicField = @sin;
    y = s.(pickField())(1); % name: indirect_access wrapping identifier -> @call
end

function name = pickField()
    name = 'dynamicField';
end

% --- command-syntax calls vs. import statement (matlab.calls.scm) -----------

function test_commands()
    disp hello % command: real call, captured as @call
    clear all % command: real call, captured as @call
    import matlab.io.* % command syntax but a language statement, NOT a call:
                        % NEGATIVE for matlab.calls.scm, POSITIVE for matlab.imports.scm
end

% --- function_definition / function_signature name variants (matlab.tags.scm)

function plain_function() % function_definition, single return via no function_output
end

function out = single_return_function() % function_definition, function_output: identifier
    out = 1;
end

function [a, b] = multi_return_function() % function_definition, function_output: multioutput_variable
    a = 1;
    b = 2;
end

classdef Interface
    methods (Abstract)
        result = mustImplement(obj) % function_signature.name: identifier -> @definition.function
    end
end

% --- class_definition.name / superclasses variants (matlab.tags.scm, matlab.types.scm)

classdef SingleParent < handle % superclasses -> property_name -> identifier: "handle"
end

classdef MultiParent < handle & matlab.mixin.Copyable
    % superclasses -> property_name (x2) -> identifier: "handle", (chained dotted name)
end

classdef NoParent
    % class_definition with no superclasses child at all: NEGATIVE for matlab.types.scm
    methods
        function obj = NoParent()
            % obj@Superclass(...) qualified call: superclass -> identifier -> @type.reference
            obj = obj@handle();
        end
    end
end

% --- complexity / nesting variants (matlab.complexity.scm) ------------------

function classify_all(n)
    if n < 0 % if_statement: @complexity, @nesting
        result = 'negative';
    elseif n == 0 % elseif_clause: @complexity (not @nesting)
        result = 'zero';
    else
        result = 'positive';
    end

    switch n % switch_statement: @complexity, @nesting
        case 1 % case_clause: @complexity
            result = 'one';
        otherwise % otherwise_clause: @complexity
            result = 'other';
    end

    for k = 1:10 % for_statement: @complexity, @nesting
        disp(k);
    end

    i = 0;
    while i < 10 % while_statement: @complexity, @nesting
        i = i + 1;
    end

    try
        risky();
    catch % catch_clause: @complexity (not its own @nesting entry)
        handle_error();
    end
end

% --- cfg.scm exit variants ---------------------------------------------------

function test_exits(n)
    for k = 1:n
        if k == 5
            break % break_statement: @cfg.exit.break
        end
        if k == 3
            continue % continue_statement: @cfg.exit.continue
        end
    end
    if n < 0
        error('bad n'); % function_call name "error": @cfg.exit.throw
    end
    if n == 0
        rethrow(lasterror()); % function_call name "rethrow": @cfg.exit.throw
    end
    result = n; % NEGATIVE: plain assignment is not an exit node
    return % return_statement: @cfg.exit.return
end

% --- NEGATIVE section: constructs that must NOT match ------------------------

function test_negatives()
    x = 5; % NEGATIVE: plain assignment, not a call, not a definition
    y = x + 1; % NEGATIVE: binary_operator, not a call
    z = notACall; % NEGATIVE: bare identifier reference, not a function_call node
    % NEGATIVE: a comment is never @call, @definition.*, or @complexity
end

File: crates/normalize-languages/tests/fixtures/zig/variants.zig
// Completeness fixture: one construct per grammar-legal variant of each
// field the zig.{tags,calls,imports,complexity,types,decorations}.scm
// queries constrain, cross-referenced against arborium-zig 2.17.0's
// node-types.json and real parse output (`normalize syntax ast` /
// `normalize syntax query`). Every construct here is *expected to be
// captured* by the relevant query; see query_fixtures.rs `zig_completeness_*`
// tests for the matrix. A NEGATIVE section at the bottom holds deliberate
// near-miss constructs that must NOT be captured, guarding against
// over-broad patterns (the root cause of every bug this pass found).

//! Container-level doc comment: decorations.scm container_doc_comment variant.

const std = @import("std");

// --- calls.scm: SuffixExpr call-site variants -------------------------------

fn plainCall() void {
    identity(); // variable_type_function: (IDENTIFIER) directly followed by FnCallArguments
}

fn dottedCall() void {
    const obj = Wrapper{};
    obj.method(); // FieldOrFnCall function_call: (IDENTIFIER) — the previously-broken shape
}

fn chainedDottedCall() void {
    const obj = Wrapper{};
    obj.inner().method(); // two FieldOrFnCall links, each its own @call
}

fn builtinCall() void {
    _ = @TypeOf(1); // BUILTINIDENTIFIER directly followed by FnCallArguments
}

const Wrapper = struct {
    pub fn method(self: @This()) void {
        _ = self;
    }
    pub fn inner(self: @This()) @This() {
        return self;
    }
};

// --- imports.scm: @import(...) call-site variants ---------------------------

const direct_import = @import("direct.zig"); // VarDecl -> @import(...)

const NestedImports = struct {
    const nested_import = @import("nested.zig"); // @import nested inside a container
};

fn importInFunction() void {
    const local_import = @import("local.zig"); // @import inside a function body
    _ = local_import;
}

// --- types.scm: type-position variants --------------------------------------

// ParamType: plain identifier
fn paramPlainType(p: Point) void {
    _ = p;
}

// ParamType: qualified/dotted identifier (FieldOrFnCall field_access leaf)
fn paramQualifiedType(a: std.mem.Allocator) void {
    _ = a;
}

// ParamType: generic instantiation (FieldOrFnCall function_call leaf) —
// Zig generics are ordinary calls, so this is structurally identical to
// paramQualifiedType except the leaf carries FnCallArguments.
fn paramGenericType(l: std.ArrayList(u8)) void {
    _ = l;
}

// ParamType wrapped in a PrefixTypeOp (slice) — the type identifier is
// still reachable; the slice prefix must not block the match.
fn paramSliceType(items: []const Point) void {
    _ = items;
}

// FnProto return type: plain identifier
fn returnsPlainType() Point {
    return Point{ .x = 0, .y = 0 };
}

// FnProto return type: qualified/dotted identifier
fn returnsQualifiedType() std.mem.Allocator {
    unreachable;
}

// VarDecl type annotation with an initializer (anchored before "=")
const typed_var_with_init: Point = Point{ .x = 1, .y = 1 };

// VarDecl type annotation with a qualified type and an initializer
const typed_qualified_var: std.mem.Allocator = undefined;

// VarDecl type annotation with NO initializer (extern decl, anchored as
// last child instead of before "=")
extern var typed_var_no_init: Point;

// ContainerField type with a default value (anchored before "=")
const HasDefaultField = struct {
    field_with_default: Point = Point{ .x = 0, .y = 0 },
};

// ContainerField type with NO default value (anchored as last child)
const NoDefaultField = struct {
    field_no_default: Point,
};

// ContainerField qualified type, no default
const QualifiedField = struct {
    alloc: std.mem.Allocator,
};

const Point = struct {
    x: f64,
    y: f64,
};

// --- tags.scm: definition variants ------------------------------------------

pub fn taggedFunction() void {} // FnProto function: (IDENTIFIER) -> @definition.function

const TaggedStruct = struct { x: i32 }; // VarDecl -> ContainerDecl (struct) -> @definition.class
const TaggedEnum = enum { a, b }; // VarDecl -> ContainerDecl (enum) -> @definition.class
const TaggedUnion = union(enum) { a: i32, b: i32 }; // VarDecl -> ContainerDecl (union) -> @definition.class

// --- complexity.scm: complexity/nesting variants ----------------------------

fn tryComplexity() !void {
    const x = try mayFail(); // UnaryExpr operator: (PrefixOp) "try" -> @complexity
    _ = x;
}

fn mayFail() !i32 {
    return 1;
}

fn catchComplexity() i32 {
    return mayFail() catch 0; // BinaryExpr operator: (BitwiseOp) "catch" -> @complexity via generic BinaryExpr
}

// --- decorations.scm: doc_comment variant -----------------------------------

/// Function-level doc comment: decorations.scm doc_comment variant.
pub fn documented() void {}

// =============================================================================
// NEGATIVE: constructs that must NOT be captured by the relevant query.
// =============================================================================

fn negativeNonCallIdentifier() void {
    const plain_value = 42; // bare VarDecl init with no ":" type annotation —
    // must NOT be captured by types.scm (no type slot exists here at all;
    // this is exactly the shape the old blanket `(IDENTIFIER) @type.reference`
    // wrongly matched).
    _ = plain_value;
}

fn negativeFieldAccessNoCall() void {
    const w = Wrapper{};
    _ = w.method; // FieldOrFnCall field_access (no FnCallArguments sibling) —
    // a bare field/method reference with no call parens must NOT be
    // captured by calls.scm's @call patterns (no FnCallArguments present).
}

fn negativeIndexingIsNotACall(items: []const i32) i32 {
    return items[0]; // SuffixOp (array indexing) — must NOT be captured by
    // calls.scm's method-call pattern; SuffixOp is unrelated to
    // FieldOrFnCall and was the prior (broken) pattern's incorrect anchor.
}

fn negativeVarDeclInitializerIsNotAType() void {
    // The initializer ErrorUnionExpr must NOT be captured as a type
    // reference even though it has the same SuffixExpr/variable_type_function
    // shape as a real type slot — it comes after "=", not after ":".
    const untyped_alias = Point;
    _ = untyped_alias;
}

fn negativeBuiltinPrimitiveTypeIsNotIdentifier(n: i32) i32 {
    // i32 is a BuildinTypeExpr keyword, not an IDENTIFIER — never captured
    // by types.scm regardless of position.
    return n;
}
